src/libAtomVM/otp_net.c

```diff
@@ -0,0 +1,322 @@
+    { ATOM_STR("\x5", "dgram"), SOCK_DGRAM },
+    { ATOM_STR("\x6", "stream"), SOCK_STREAM },
+    SELECT_INT_DEFAULT(UNKNOWN_TABLE_VALUE)
+};
+
+//
+// utilities
+//
+
+static term socket_tuple_from_addr(Context *ctx, uint32_t addr)
+{
+    term terms[4];
+    terms[0] = term_from_int32((addr >> 24) & 0xFF);
+    terms[1] = term_from_int32((addr >> 16) & 0xFF);
+    terms[2] = term_from_int32((addr >> 8) & 0xFF);
+    terms[3] = term_from_int32(addr & 0xFF);
+
+    return port_create_tuple_n(ctx, 4, terms);
+}
+
+static inline term make_error_tuple(term reason, Context *ctx)
+{
+    term error_tuple = term_alloc_tuple(2, &ctx->heap);
+    term_put_tuple_element(error_tuple, 0, ERROR_ATOM);
+    term_put_tuple_element(error_tuple, 1, reason);
+    return error_tuple;
+}
+
+static term eai_errno_to_term(int err, GlobalContext *glb)
+{
+    switch (err) {
+        case EAI_AGAIN:
+            return globalcontext_make_atom(glb, ATOM_STR("\x8", "eaiagain"));
+        case EAI_BADFLAGS:
+            return globalcontext_make_atom(glb, ATOM_STR("\xB", "eaibadflags"));
+        case EAI_FAIL:
+            return globalcontext_make_atom(glb, ATOM_STR("\x7", "eaifail"));
+        case EAI_FAMILY:
+            return globalcontext_make_atom(glb, ATOM_STR("\x9", "eaifamily"));
+        case EAI_MEMORY:
+            return globalcontext_make_atom(glb, ATOM_STR("\x9", "eaimemory"));
+        case EAI_NONAME:
+            return globalcontext_make_atom(glb, ATOM_STR("\x9", "eainoname"));
+        case EAI_SERVICE:
+            return globalcontext_make_atom(glb, ATOM_STR("\xA", "eaiservice"));
+        case EAI_SOCKTYPE:
+            return globalcontext_make_atom(glb, ATOM_STR("\xB", "eaisocktype"));
+
+            // These constants are not supported on ESP32
+            // case EAI_BADHINTS:
+            //     return globalcontext_make_atom(glb, ATOM_STR("\xB", "eaibadhints"));
+            // case EAI_OVERFLOW:
+            //     return globalcontext_make_atom(glb, ATOM_STR("\xB", "eaioverflow"));
+            // case EAI_PROTOCOL:
+            //     return globalcontext_make_atom(glb, ATOM_STR("\xB", "eaiprotocol"));
+            // case EAI_SYSTEM:
+            //     return globalcontext_make_atom(glb, ATOM_STR("\x9", "eaisystem"));
+    }
+    return term_from_int(err);
+}
+
+static inline term make_address(Context *ctx, struct sockaddr_in *addr)
+{
+    return socket_tuple_from_addr(ctx, ntohl(addr->sin_addr.s_addr));
+}
+
+//
+// net:getaddrinfo/1
+//
+
+static term nif_net_getaddrinfo(Context *ctx, int argc, term argv[])
+{
+    TRACE("nif_net_getaddrinfo\n");
+    UNUSED(argc);
+
+    GlobalContext *global = ctx->global;
+
+    term host = argv[0];
+    term service = argv[1];
+
+    if (host == UNDEFINED_ATOM && service == UNDEFINED_ATOM) {
+        TRACE("Host and Service params may not both be undefined");
+        RAISE_ERROR(BADARG_ATOM);
+    }
+
+    if (!(term_is_string(host) || host == UNDEFINED_ATOM)) {
```

Comment:
Because we are parsing the list below with interop_term_to_string, we should only do term_is_list here.

Reply:
Have a look at the change